docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,245 @@
+Requirements](https://github.com/cosmos/ics/tree/master/spec/ics-024-host-requirements#paths-identifiers-separators)
+specification.
+
+## Decision
+
+Instead of adding the identifiers on the coin denomination directly, the proposed solution hashes the denomination prefix in order to get a consistent lenght for all the cross-chain fungible tokens. The new format will be the following:
+
+```golang
+ibcDenom = "ibc/" + SHA256 hash of the trace identifiers prefix + "/" + base coin denomination
+```
+
+### `x/ibc-transfer` Changes
+
+In order to retreive the trace information from an IBC denomination, a lookup table needs to be added to the `ibc-transfer` module. These values need to also be persisted between upgrades, meaning that a new `[]Trace` `GenesisState` field state needs to be added to the module:
+
+```golang
+// GetDenom retreives the full identifiers trace from the store.
+func (k Keeper) GetTrace(ctx Context, traceHash []byte) string {
+  store := ctx.KVStore(k.storeKey)
+  bz := store.Get(types.KeyTrace(traceHash))
+  if len(bz) == 0 {
+    return ""
+  }
+  return string(bz)
+}
+
+// HasTrace checks if a the key with the given trace hash exists on the store.
+func (k Keeper) HasTrace(ctx Context, traceHash []byte)  bool {
+  store := ctx.KVStore(k.storeKey)
+  return store.Has(types.KeyTrace(traceHash))
+}
+
+// SetTrace sets a new {trace hash -> trace} pair to the store.
+func (k Keeper) SetTrace(ctx Context, traceHash []byte, trace string) {
+  store := ctx.KVStore(k.storeKey)
+  store.Set(types.KeyTrace(traceHash), []byte(trace))
+}
+```
+
+When a fungible token with a is send to a sink chain, the trace information needs to be updated with the new port and channel identifiers:
+
+```golang
+func (k Keeper) PrefixDenom(ctx Context, portID, channelID, denom string) string {
+  // Get each component of the denom. The resulting slice will be:
+  //
+  // - [ "ibc", traceHash, baseDenom], if the denom is dirty (contains trace metadata).
+  // - [ baseDenom ], if the denom has never been sent from the origin chain.
+  denomSplit := strings.Split(denom, "/")
+
+  var (
+    baseDenom string
+    trace     string
+    traceHash tmbytes.HexBytes
+  )
+
+  // check if the denomination is clean or if it contains the trace info
+  if denomSplit[0] == denom {
+    baseDenom = denom
+    trace = portID + "/" + channelID +"/"
+  } else {
+    baseDenom = denomSplit[2]
+    traceHash = tmbytes.HexBytes(denomSplit[1])
+    // Get the value from the map trace hash -> denom identifiers prefix
+    trace = k.GetTrace(ctx, traceHash)
+    // prefix the identifiers to create the new trace
+    trace = portID + "/" + channelID +"/" + trace + "/"
+  }
+  
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String() + baseDenom
```

Comment:
This is *not* what is mentioned above, I believe:

above:

```golang
ibcDenom = "ibc/" + SHA256 hash of the trace identifiers prefix + "/" + base coin denomination
```

here:

```golang
ibcDenom = "ibc/" + SHA256 hash of (the trace identifiers prefix + "/") + base coin denomination
```

I think the former is what you wanted?

Reply:
good catch!